kotlinx-coroutines-core/jvm/test/sync/SemaphoreStressTest.kt

```diff
@@ -66,35 +59,48 @@ class SemaphoreStressTest : TestBase() {
         assertEquals(1, semaphore.availablePermits)
     }
 
+    /* This checks if repeated releases that race with cancellations put
+       the semaphore into an incorrect state where permits are leaked. */
     @Test
     fun stressReleaseCancellation() = runTest {
         val n = 10_000 * stressTestMultiplier
         val semaphore = Semaphore(1, 1)
-        repeat (n) {
-            assertEquals(0, semaphore.availablePermits)
 
-            var shared = false
+        newSingleThreadContext("SemaphoreStressTest").use { pool ->
 
-            val job1 = launch(pool) {
-                semaphore.acquire()
-                shared = true
-                semaphore.release()
-            }
+            repeat (n) {
+                /* Initially, we hold the permit and no one else can `acquire`,
+                   otherwise it's a bug. */
+                assertEquals(0, semaphore.availablePermits)
 
-            assertEquals(false, shared)
+                var job1_entered_critical_section = false
 
-            val job2 = launch(pool) {
-                semaphore.release()
-            }
+                val job1 = launch {
+                    semaphore.acquire()
+                    job1_entered_critical_section = true
+                    semaphore.release()
+                }
 
-            job1.cancelAndJoin()
-            job2.join()
+                // Let `job1` run and sleep in `acquire`
+                yield()
 
-            assertEquals(1, semaphore.availablePermits)
-            semaphore.acquire()
-        }
+                assertEquals(false, job1_entered_critical_section)
 
-        pool.close()
+                val job2 = launch(pool) {
+                    semaphore.release()
+                }
+
```

Comment:
style: try to avoid extra blank lines inside the code functions. Longer vertical size makes it harder to understand. Separate logical blocks of code with comments or not separate them at all.